File: BotManager/BotLib/src/main/java/com/rkarp/botlib/runner/BotRunner.java
package com.rkarp.botlib.runner;

import com.rkarp.botlib.model.Credentials;
import javafx.scene.Node;
import javafx.scene.control.Button;

/**
 * This class starts the bot on the defined platform.
 * Supported browsers: chrome, firefox
 * Supported mobile platforms: android
 * @param <DRI> Driver
 * @param <SVC> Driver service
 */
public abstract class BotRunner<DRI, SVC> {
    private DRI object;
    public DRI get() {
        return object;
    }
    void set(DRI obj) {
        this.object = obj;
    }

    private SVC driverService;
    SVC getDriverService() {
        return driverService;
    }
    public void setDriverService(SVC obj) {
        this.driverService = obj;
    }

    public abstract void start();

    public abstract void setLayout(Node layoutNode);

    public abstract void addSetting(String title, String description, boolean isVisible, Node settingNode);

    public abstract void addControlButton(Button... controlButton);
    public abstract void removeControlButton(int index);


    //--- Credentials
    private Credentials loginCredentials;
    public Credentials getCredentials() {
        return loginCredentials;
    }
    public void setCredentials(Credentials obj) {
        this.loginCredentials = obj;
    }
    public boolean hasCredentials() {
        return loginCredentials != null;
    }
}

File: projects/JavaFXApps/BotManager/BotLib/src/main/java/net/rickiekarp/botlib/runner/BotRunner.kt
package net.rickiekarp.botlib.runner

import net.rickiekarp.botlib.model.Credentials
import javafx.scene.Node
import javafx.scene.control.Button

/**
 * This class starts the bot on the defined platform.
 * Supported browsers: chrome, firefox
 * Supported mobile platforms: android
 * @param <DRI> Driver
 * @param <SVC> Driver service
</SVC></DRI> */
abstract class BotRunner<DRI, SVC> {
    private var `object`: DRI? = null

    var driverService: SVC? = null


    //--- Credentials
    var credentials: Credentials? = null

    fun get(): DRI? {
        return `object`
    }

    internal fun set(obj: DRI) {
        this.`object` = obj
    }

    abstract fun start()

    abstract fun setLayout(layoutNode: Node)

    abstract fun addSetting(title: String, description: String, isVisible: Boolean, settingNode: Node)

    abstract fun addControlButton(vararg controlButton: Button)
    abstract fun removeControlButton(index: Int)
    fun hasCredentials(): Boolean {
        return credentials != null
    }
}
